[JSC] Add lowest/highest accessible address functions

Constellation opened the pull request:
#### 1c884ddcfc16ea40d821dd4ed28594860d7d51df
<pre>
[JSC] Add lowest/highest accessible address functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=249984">https://bugs.webkit.org/show_bug.cgi?id=249984</a>
rdar://103800354

Reviewed by Mark Lam.

This patch adds lowestAccessibleAddress / highestAccessibleAddress.
On Darwin, lowestAccessibleAddress gets __PAGEZERO information of process&apos; executable.

* Source/JavaScriptCore/tools/Integrity.h:
(JSC::Integrity::isSanePointer):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/AccessibleAddress.cpp: Added.
(WTF::lowestAccessibleAddress):
(WTF::highestAccessibleAddress):
* Source/WTF/wtf/AccessibleAddress.h: Added.
* Source/WTF/wtf/CMakeLists.txt:

Canonical link: <a href="https://commits.webkit.org/258376@main">https://commits.webkit.org/258376@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1077c9c4a0ddb408eae697749144a8d5993f40b0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101723 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10873 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34795 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111057 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/171262 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/105703 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11830 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1785 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94131 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108819 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107503 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9042 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92301 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/36693 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90934 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23728 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/78611 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/92144 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4471 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25220 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/88286 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/2069 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4546 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/1668 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29377 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10634 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44707 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/91193 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6300 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20369 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3028 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/WTF/wtf/AccessibleAddress.cpp:
```diff
@@ -0,0 +1,80 @@
+{
+    static uintptr_t cached = 0;
+    static std::once_flag onceKey;
+    std::call_once(onceKey, []() {
```
Instead of using a `static` field and doing a `once_flag` check, can you put this field in `WTF::Config` and create an `initializeAccessibleAddress` to be called from `WTF::initialize` (in Threading.cpp) instead?  That way, `lowestAccessibleAddress()` can be an inline function.  `lowestAccessibleAddress()` can also ASSERT that the field in `WTF::Config` is initialized (i.e. not null).  I think we want `isSanePointer` to be cheap.

Maybe we should make `highestAccessibleAddress` an inline function too.

Constellation replied:
That's nice! Changed.

MenloDorian reviewed (approved):
r=me with suggestion.

Constellation committed:
[JSC] Add lowest/highest accessible address functions
https://bugs.webkit.org/show_bug.cgi?id=249984
rdar://103800354

Reviewed by Mark Lam.

This patch adds lowestAccessibleAddress / highestAccessibleAddress.
On Darwin, lowestAccessibleAddress gets __PAGEZERO information of process' executable.

* Source/JavaScriptCore/tools/Integrity.h:
(JSC::Integrity::isSanePointer):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/AccessibleAddress.cpp: Added.
(WTF::lowestAccessibleAddress):
(WTF::highestAccessibleAddress):
* Source/WTF/wtf/AccessibleAddress.h: Added.
* Source/WTF/wtf/CMakeLists.txt:

Canonical link: https://commits.webkit.org/258376@main
